### PR TITLE
[Dashboard] Dashboard basic modules

### LINKED ISSUE #10016: [Dashboard] Port reporter.py and related tests to the new dashboard
1. Port `reporter.py` to `reporter_agent.py` in `reporter` module.
2. Port RESTful APIs of reporter to `reporter_head.py` in `reporter` module.
3. Port related tests to `reporter` module.

### LINKED ISSUE #10017: [Dashboard] A log module for the new dashboard
- Each agent starts a http log agent to provide log access.
- The log head provides endpoints of log agents.

### LINKED ISSUE #10019: [Dashboard] A stats_collector module for collecting node info
- **GET** `/nodes?view=summary|hostNameList`
- **GET** `/node/{hostname}`
- `stats_collector` also collects actors.

### PR DESCRIPTION
**Pull request by fyrestone**
This PR includes 3 basic modules for new dashboard:
- reporter - Forks `reporter.py`, the profiling RESTful API has been simplified to one `GET /api/launch_profiling`.
- log - Basic log module, provides `GET /log_index`, `GET /logs` from dashboard and `GET /logs` from dashboard agent.
- stats_collector - Collects stats (currently, actors and node_stats), provides `GET /nodes?view=[summary|hostNameList]` and `GET /nodes/{hostname}`

## Related issue number

- Closes #10016
- Closes #10017
- Closes #10019

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: dashboard/head.py
Comment (rkooo567) [inline dashboard/head.py:81]: Can you explain the reason in the comment too?

File: dashboard/modules/log/log_head.py
Comment (rkooo567) [inline dashboard/modules/log/log_head.py:32]: Can you just use fstring? We are trying to move onto fstring in general. I think it is fine not to use template because it is used only for 3 places.

File: dashboard/modules/log/log_head.py
Comment (rkooo567) [inline dashboard/modules/log/log_head.py:54]: Please use fstring here too.

File: dashboard/modules/log/test_log.py
Comment (rkooo567) [inline dashboard/modules/log/test_log.py:48]: Can you reduce it to 10 seconds? 10 seconds timeout makes test flaky we should break them down instead.

File: dashboard/modules/reporter/reporter_agent.py
Comment (rkooo567) [inline dashboard/modules/reporter/reporter_agent.py:105]: Nice. Thanks for adding this back.

File: dashboard/modules/reporter/reporter_head.py
Comment (rkooo567) [inline dashboard/modules/reporter/reporter_head.py:46]: Is it intended?

File: dashboard/modules/reporter/reporter_head.py
Comment (rkooo567) [inline dashboard/modules/reporter/reporter_head.py:51]:
```suggestion
                if reply.profiling_stats else reply.std_out)
```

File: dashboard/modules/reporter/test_reporter.py
Comment (rkooo567) [inline dashboard/modules/reporter/test_reporter.py:38]: Reduce timeout to 10 seconds here too!

File: dashboard/tests/test_dashboard.py
Comment (rkooo567) [inline dashboard/tests/test_dashboard.py:27]: Why do we need this?

**Review by rkooo567 — COMMENTED** [review]
It LGTM in general (and nice addition of tests :)), but since the PR is really large, I will have another batch of review later. Next time when you submit a PR, please break them down (this could be broken down to 3 PRs imo).

File: dashboard/head.py
Comment (fyrestone) [reply dashboard/head.py:81]: `GetAllNodeInfo` returns all nodes info including dead nodes. We needs to know how to merge them to ip. For example, there are two nodes info:
- {node id: 1, hostname: example}
- {node id: 2, hostname: example}

We choose which one for host example?

For one hostname, there will be a list of node info. and only two cases:
- All nodes info of one hostname are DEAD
- Only one node info of one hostname is ALIVE

So, here is the rule,
- Choose a DEAD one if all nodes info of one hostname are DEAD. 
- Choose the ALIVE one if there is ALIVE node in one hostname.

Comment (fyrestone) [reply dashboard/head.py:81]: It's better to add a timestamp to `GcsNodeInfo`.

File: dashboard/modules/log/test_log.py
Comment (fyrestone) [reply dashboard/modules/log/test_log.py:48]: OK. 10 seconds is enough.

File: dashboard/modules/reporter/reporter_head.py
Comment (fyrestone) [reply dashboard/modules/reporter/reporter_head.py:46]: No. I will remove this line.

File: dashboard/modules/reporter/test_reporter.py
Comment (fyrestone) [reply dashboard/modules/reporter/test_reporter.py:38]: I will reduce all the timeout to 10 seconds.

File: dashboard/tests/test_dashboard.py
Comment (fyrestone) [reply dashboard/tests/test_dashboard.py:27]: This is for `test_async_loop_forever`.

File: dashboard/modules/log/log_head.py
Comment (fyrestone) [reply dashboard/modules/log/log_head.py:32]: So, `ray` has dropped support for Python 3 - 3.6? I prefer to use fstring than old formatter.

File: dashboard/head.py
Comment (rkooo567) [reply dashboard/head.py:81]: Hmm.. I see. So this is due to the case where there could be multiple raylets on a single host. I think this should be the case only when we use `cluster_utils` right? Or do you know any other case? I think it kind of doesn't make sense to have multiple ray start in a single node. 

If this is useful only for the cluster utils, why don't we just group by hostname + node_id and display both in the dashboard? Is there any issue with this?

File: dashboard/modules/log/log_head.py
Comment (rkooo567) [reply dashboard/modules/log/log_head.py:32]: It dropped the support for 3-3.5. Afaik, 3.6 supports fstring?

File: dashboard/tests/test_dashboard.py
Comment (rkooo567) [inline dashboard/tests/test_dashboard.py:335]: 4 seconds delay might be too long?

File: dashboard/tests/test_dashboard.py
Comment (rkooo567) [reply dashboard/tests/test_dashboard.py:27]: Hmm.. I am still confused. Can you elaborate a little more and add comments too? Is there any way we can remove this at all?

File: dashboard/modules/log/log_head.py
Comment (fyrestone) [reply dashboard/modules/log/log_head.py:54]: I will use fstring instead.

File: dashboard/modules/reporter/reporter_head.py
Comment (fyrestone) [reply dashboard/modules/reporter/reporter_head.py:51]: I will assign the conditional expression to a variable.

File: dashboard/head.py
Comment (fyrestone) [reply dashboard/head.py:81]: If a node failover many times, there will be a lot node info belongs to one hostname.

File: dashboard/modules/log/log_head.py
Comment (fyrestone) [reply dashboard/modules/log/log_head.py:32]: Yes, Python 3.6 supports fstring. I am happy to know that I can use fstring.

File: dashboard/tests/test_dashboard.py
Comment (fyrestone) [reply dashboard/tests/test_dashboard.py:335]: It's better to set `interval_seconds` to 0.1 and `call_later` 1 seconds. I will fix it.

File: dashboard/head.py
Comment (rkooo567) [reply dashboard/head.py:81]: I see. This hostname + node id idea still should work right? (We can just ignore all DEAD nodes). Is there any concern about this approach?

Comment (fyrestone) [reply dashboard/head.py:81]: If a node failover frequently, `hostname + node id` always get a DEAD node info. I think it's better to query the node info by hostname and show the history by DEAD node info. For example,

- {node id: 1, start_timestamp: 10, dead timestamp: 100, state: DEAD, hostname: example}
- {node id: 2, start_timestamp: 102, dead timestamp: 120, state: DEAD, hostname: example}
- {node id: 3, start_timestamp: 122, dead timestamp: 0, state: ALIVE, hostname: example}

then the node `example` will show the third node info with history:
- node id: 2, start at: 102, dead at: 120
- node id: 1 start at: 10, dead at: 100

File: dashboard/tests/test_dashboard.py
Comment (fyrestone) [reply dashboard/tests/test_dashboard.py:27]: Sorry, `test_async_loop_forever` use `loop.create_task` not `asyncio.create_task`. This 4 lines can be removed, I will fix it.

File: dashboard/modules/reporter/reporter_agent.py
Comment (fyrestone) [reply dashboard/modules/reporter/reporter_agent.py:105]: I hope we can move to the new dashboard as soon as possible.

File: dashboard/modules/stats_collector/stats_collector_head.py
Comment (mfitton) [inline dashboard/modules/stats_collector/stats_collector_head.py:146]: Could you add a bit more information to the error message here? logger.exception will already capture the exception in its body so you could do 
```
except Exception:
    logger.exception(f"Error fetching node stats for node with ip={ip}")
```
https://docs.python.org/3/library/logging.html#logging.Logger.exception

File: dashboard/modules/stats_collector/stats_collector_head.py
Comment (mfitton) [inline dashboard/modules/stats_collector/stats_collector_head.py:131]: See similar comment below about logger.exception

File: dashboard/modules/stats_collector/stats_collector_head.py
Comment (mfitton) [inline dashboard/modules/stats_collector/stats_collector_head.py:117]: See comment below about logger.exception

File: dashboard/tests/test_dashboard.py
Comment (mfitton) [inline dashboard/tests/test_dashboard.py:60]: Why does this need to be disabled?

File: dashboard/tests/conftest.py
Comment (mfitton) [inline dashboard/tests/conftest.py:7]: One thing you could also do here is change this to a statement that cleans itself up. For instance:
```
@pytest.fixture(scope="module")
def disable_test_module():
    os.environ.pop("RAY_DASHBOARD_MODULE_TEST", None)
    yield
    os.environ["RAY_DASHBOARD_MODULE_TEST"] = "true"
```
which would automatically re-enable tests after the module exits.

Then I think you would only need one of these two fixtures.

File: dashboard/utils.py
Comment (mfitton) [inline dashboard/utils.py:176]: Nice, this is clean.

File: dashboard/utils.py
Comment (mfitton) [inline dashboard/utils.py:411]: See other comment about logger.exception. You can leave body blank or add more details.

**Review by mfitton — COMMENTED** [review]
Overall, this looks really good. I had a couple questions, and a couple other pieces of minor feedback, but hopefully it'll be quick to address. Thanks for all your hard work on this @fyrestone 
Happy to approve once you address or reply to my comments.

File: dashboard/head.py
Comment (rkooo567) [reply dashboard/head.py:81]: > If a node failover frequently, hostname + node id always get a DEAD node info.
In this case, we can just not render them right? We can use your logic in the frontend.

The reason why I want to use these as keys is because it'll be great if we can work on dashboard with cluster_utils. The current structure won't work with cluster_utils because it will have the same hostname for multiple node ids.

File: dashboard/agent.py
Comment (fyrestone) [reply dashboard/agent.py:72]: The best formatter for logger is `%`, because we may set different log levels, and `%` formatter will avoid formatting skipped levels. For example, we set log level to `ERROR`, then
```python
logger.info("Loaded %d modules.", len(modules))  # Skip format.
logger.info("Loaded {} modules.".format(len(modules)))  # Always format.
```

File: dashboard/dashboard.py
Comment (fyrestone) [reply dashboard/dashboard.py:40]: Thanks to fstring for making the code clean.

File: dashboard/agent.py
Comment (fyrestone) [reply dashboard/agent.py:30]: Yes, the `asyncio.ensure_future` is the older version of `asyncio.create_task`. From the official doc of `asyncio.create_task` https://docs.python.org/3/library/asyncio-task.html#asyncio.create_task:
```
This function has been added in Python 3.7. Prior to Python 3.7,
the low-level asyncio.ensure_future() function can be used instead
```

File: dashboard/tests/test_dashboard.py
Comment (fyrestone) [reply dashboard/tests/test_dashboard.py:60]: Test whether the basic functions are good when the test module is disabled.

File: dashboard/head.py
Comment (fyrestone) [reply dashboard/head.py:81]: @mfitton @rkooo567 We want to find the best solution between (IP, hostname) and (node id).

- For (IP, hostname)
  - Pros: simple for users, friendly to one host one node. (It's easy to show the host history)
  - Cons: not support one host multiple nodes, so it's not work well with cluster_utils.

- For (node id)
  - Pros: support one host multiple nodes, work well with cluster_utils.
  - Cons: too many nodes info may confuse users, hard to find out what happened if node failover many times.

Comment (rkooo567) [reply dashboard/head.py:81]: I think we can hide the second options' con easily by filtering DEAD node ids to user facing APIs right?

Comment (rkooo567) [reply dashboard/head.py:81]: {
node_id: {
  ip:
  host:
  state:
}
}

And in the frontend, we can just filter DEAD state node id.

File: dashboard/agent.py
Comment (rkooo567) [reply dashboard/agent.py:30]: Why not just always using ensure_future in this case?

File: dashboard/modules/log/log_head.py
Comment (fyrestone) [reply dashboard/modules/log/log_head.py:50]: This function is to show the node log entries by html from log head, the html structure is same with show_index of static files from log agent. The log structure is:
- /log_index (`_directory_as_html` logic shows the html from log head)
  - /logs (static files with `show_index=True` shows the html from log agent)
    - dashboard.log 
    - dashboard_agent.log
  - /logs (static files with `show_index=True` shows the html from log agent)
    - dashboard.log
 ...

@mxz96102 will provides the front-end of log module with advanced features.

File: dashboard/modules/reporter/reporter_head.py
Comment (fyrestone) [reply dashboard/modules/reporter/reporter_head.py:59]: Thanks.

File: dashboard/modules/reporter/reporter_head.py
Comment (fyrestone) [reply dashboard/modules/reporter/reporter_head.py:40]: Thanks. `KeyError` is clearer than `TypeError`. I will refine the logic.

File: dashboard/modules/stats_collector/stats_collector_head.py
Comment (fyrestone) [reply dashboard/modules/stats_collector/stats_collector_head.py:146]: Good suggestion. `logger.exception(ex)` is to show the message of `ex` with traceback. I will refine the message with details.

File: dashboard/agent.py
Comment (fyrestone) [reply dashboard/agent.py:30]: Because Python doc recommends using `asyncio.create_task`. see: https://docs.python.org/3/library/asyncio-future.html#asyncio.ensure_future
```
Important See also the create_task() function which is the preferred way for creating new Tasks.
```

File: dashboard/head.py
Comment (fyrestone) [reply dashboard/head.py:81]: @rkooo567 I see. Use node id as the key of node info will returns full nodes info to front-end. And front-end do the filter job. My concern is that the data will be too large if we run many nodes in one cluster. For the DEAD node, the physics node info is useless. If we choose `node id` as the key, we need some strategies to reduce the data size:
- Only contains `GcsNodeInfo` data for the DEAD nodes.
- Each (IP + hostname) reserves a limited number of DEAD nodes. For example, the last 5 DEAD nodes.

File: dashboard/tests/conftest.py
Comment (fyrestone) [reply dashboard/tests/conftest.py:7]: Good idea. The test module is disabled by default, so one `enable_test_module()` is enough for the tests. I will merge the `disable_test_module()` and `enable_test_module()`.

File: dashboard/modules/stats_collector/stats_collector_head.py
Comment (fyrestone) [reply dashboard/modules/stats_collector/stats_collector_head.py:56]: I will add a simple type for the Dict change.

File: dashboard/utils.py
Comment (fyrestone) [reply dashboard/utils.py]: Fixed.

File: dashboard/utils.py
Comment (fyrestone) [reply dashboard/utils.py:177]: Thanks.

File: dashboard/modules/stats_collector/stats_collector_head.py
Comment (fyrestone) [reply dashboard/modules/stats_collector/stats_collector_head.py]: Fixed.

File: dashboard/modules/stats_collector/stats_collector_head.py
Comment (fyrestone) [reply dashboard/modules/stats_collector/stats_collector_head.py]: Fixed.

File: dashboard/agent.py
Comment (mfitton) [reply dashboard/agent.py:72]: That's a good point that I hadn't thought of. Thanks!

File: dashboard/head.py
Comment (rkooo567) [reply dashboard/head.py:81]: I see your concerns (since you guys are running long-running clusters, it can have lots of information). I like the second solution, but we can allow users to configure them. So, for regular usage, it is 0~2, and for cluster_utils cases, it can be like 10. 

> Only contains GcsNodeInfo data for the DEAD nodes.

For this, I am probably not familiar how the data is represented now (because I thought this was the current behavior). Can you explain a little bit?

**Review by rkooo567 — APPROVED** [review]
I think the only concern I have is how we will group nodes. I prefer to make it support cluster_utils because we will have more distributed data collection we'd like to test in the future (for example, we will remove all GetCoreWorkerStats, and replace this with our metrics infra. This is hard to be tested without cluster_utils). 

But, I think this is not a hard blocker for this PR. (but I think this should be the next dashboard task).

**Comment by rkooo567** [comment]
Maybe another solution is to just refactor our node.py to contain hostname. In that way, we can inject the fake hostname for clutser_utils.

File: dashboard/head.py
Comment (fyrestone) [reply dashboard/head.py:81]: Current node info is a mixture of `node physical stats` (from **_reporter agent_**) and `node stats` (from **_GetNodeStats_** rpc). If a node is dead, the `node physical stats` and `node stats` will be unreliable, only `GcsNodeInfo` is correct.

**Comment by fyrestone** [comment]
> I think the only concern I have is how we will group nodes. I prefer to make it support cluster_utils because we will have more distributed data collection we'd like to test in the future (for example, we will remove all GetCoreWorkerStats, and replace this with our metrics infra. This is hard to be tested without cluster_utils).
> 
> But, I think this is not a hard blocker for this PR. (but I think this should be the next dashboard task).

I agree with you. I can make a PR after we find a solution.

**Review by mfitton — APPROVED** [review]

**Review by mfitton — APPROVED** [review]
Looks good. It seems to me like the test failures are unrelated, and I agree with Sang that the PR shouldn't be blocked by the node grouping problem.

**Comment by rkooo567** [comment]
@fyrestone Please resolve the merge conflict and tag `test_Ok` label once build result looks fine. I will merge it after that :)

**Comment by fyrestone** [comment]
> @fyrestone Please resolve the merge conflict and tag `test_Ok` label once build result looks fine. I will merge it after that :)

I have resolved the conflicts, and set the `test_Ok` label. I am waiting for the test results.

**Comment by rkooo567** [comment]
Thanks again for this PR @fyrestone! It looks really clean, and I am so happy about testing status. Would you mind pushing a PR for the node grouping next? (if you are planning to work on that)!

**Comment by fyrestone** [comment]
> Thanks again for this PR @fyrestone! It looks really clean, and I am so happy about testing status. Would you mind pushing a PR for the node grouping next? (if you are planning to work on that)!

Thanks. I will create the node grouping PR this week. The `GET /nodes` will return all nodes info using `node id` as the key, and `GET /node/{hostname}` will be replaced by `GET /node/{nodeid}`.

**Pull request merged**
